File: samples/java/src/main/java/org/jrosenkranz/opts/java_api/samples/OptsSample.java
package org.jrosenkranz.opts.java_api.samples;

import org.jrosenkranz.opts.java_api.Opts;
import org.jrosenkranz.opts.java_api.options.OptBuilder;
import org.jrosenkranz.opts.java_api.set.OptSet;

import java.util.Arrays;
import java.util.List;

public class OptsSample {

    private static OptSet optSet = OptSet.build(
            OptBuilder.defaultOpt("topics", Arrays.asList("sample1","sample2","sample3"))
                    .message("kafka topics comma separated")
                    .parseOp(s -> Arrays.asList(s.split(",")))
                    .abrev("t"),
            OptBuilder.defaultOpt("inputPath","./").message("path to input"),
            OptBuilder.defaultOpt("delimiter",",").message("kafka message delimiter"),
            OptBuilder.flagOpt("help").message("this will display all commands as a list"),
            OptBuilder.commandOpt("update").message("update the system")
                    .addChild(OptBuilder.requiredOpt("host").message("kafka host"))
                    .addChild(OptBuilder.defaultOpt("port",2181).message("kafka port").parseOp(Integer::valueOf))
                    .addChild(OptBuilder.flagOpt("help").message("this will display the update help")),
            OptBuilder.commandOpt("delete").message("delete the system")
                    .addChild(OptBuilder.requiredOpt("name").message("item to delete"))
    );

    public static void main(String... args) {
        Opts options = new Opts(optSet,args);

        if (options.<Boolean>getAs("help")) {
            System.out.println(optSet);
            System.exit(0);
        }

        List<String> topics = options.getAs("topics");
        String inputPath = options.getAs("inputPath");
        String delimiter = options.getAs("delimiter");
        topics.forEach(System.out::println);
        System.out.println(inputPath);
        System.out.println(delimiter);

        if (options.commands.isPresent()) {

            switch (options.commands.get().get(0)) {
                case "update":
                    String host = options.getAs("host");
                    int port = options.getAs("port");
                    boolean help = options.getAs("help");
                    System.out.println(host);
                    System.out.println(port);
                    System.out.println(help);
                    break;

                case "delete":
                    String name = options.getAs("name");
                    System.out.println(name);
                    break;
            }


        }
    }
}

File: samples/scala/src/main/scala/org/jrosenkranz/opts/samples/OptsSample.scala
package org.jrosenkranz.opts.samples

import org.jrosenkranz.opts.Opts
import org.jrosenkranz.opts.options.{CommandOpt, DefaultOpt, FlagOpt, RequiredOpt}
import org.jrosenkranz.opts.set.{ChildOptSet, OptSet}

object OptsSample {
  val os = OptSet(
    DefaultOpt("topics",List("sample"), "kafka topics comma separated","t",_.split(",").toList),
    DefaultOpt("inputPath","./","path to input"),
    DefaultOpt("delimiter",",","kafka message delimiter"),
    FlagOpt("help","this will display all commands as a list"),
    CommandOpt("update","update the system",ChildOptSet(
      RequiredOpt("host","kafka host"),
      DefaultOpt("port",2181,"kafka port"),
      FlagOpt("help","this will display the update help")
    )),
    CommandOpt("delete","delete the system",ChildOptSet(
      RequiredOpt("name","item to delete")
    ))
  )

  def main(args: Array[String]): Unit = {

    os.parse(args) match {
      //parsing passed
      case Some(options) => {

        //check what to do depending on the commands given
        options.commands match {

          case Some(Array("update")) => handleUpdate(options)

          case Some(Array("delete")) => handleDelete(options)

          case None => handleNoCommand(options)
        }

      }
      //parsing failed
      case None => println(os.toString())
    }
  }

  private def handleUpdate(options: Opts): Unit = {
    println(options)
  }

  private def handleDelete(options: Opts): Unit = {
    println(options)
  }

  private def handleNoCommand(options: Opts): Unit = {
    //if help is true, print out the available commands
    if (options.as[Boolean]("help")) {
      println(os.filter(_.isCmd).map(opt => s"Command: ${opt.key}   ${opt.message}").mkString("\n"))
    } else {
      println(os)
    }
  }
}
